File: test_vs/05_agent_pipeline.py
"""Benchmark: Agent Pipeline Processing Messages — Python asyncio baseline."""

import asyncio
import time


async def parser(inbox: asyncio.Queue, outbox: asyncio.Queue, stop: asyncio.Event):
    while not stop.is_set():
        try:
            raw = await asyncio.wait_for(inbox.get(), timeout=0.1)
            await outbox.put(len(raw))
        except asyncio.TimeoutError:
            continue


async def validator(inbox: asyncio.Queue, outbox: asyncio.Queue, stop: asyncio.Event):
    while not stop.is_set():
        try:
            value = await asyncio.wait_for(inbox.get(), timeout=0.1)
            await outbox.put(value if value > 0 else 0)
        except asyncio.TimeoutError:
            continue


async def transformer(inbox: asyncio.Queue, outbox: asyncio.Queue, stop: asyncio.Event):
    while not stop.is_set():
        try:
            value = await asyncio.wait_for(inbox.get(), timeout=0.1)
            await outbox.put(value * 3 + 7)
        except asyncio.TimeoutError:
            continue


async def main():
    p_in: asyncio.Queue = asyncio.Queue(maxsize=1024)
    p_out: asyncio.Queue = asyncio.Queue(maxsize=1024)
    v_out: asyncio.Queue = asyncio.Queue(maxsize=1024)
    t_out: asyncio.Queue = asyncio.Queue(maxsize=1024)

    stop = asyncio.Event()

    t1 = asyncio.create_task(parser(p_in, p_out, stop))
    t2 = asyncio.create_task(validator(p_out, v_out, stop))
    t3 = asyncio.create_task(transformer(v_out, t_out, stop))

    start = time.perf_counter()
    total = 0
    for i in range(1000):
        msg = "message_payload_data_" + str(i)
        await p_in.put(msg)
        result = await t_out.get()
        total += result
    elapsed = time.perf_counter() - start

    stop.set()
    t1.cancel()
    t2.cancel()
    t3.cancel()

    print(f"sum = {total}")
    print(f"Time: {elapsed:.4f}s")


if __name__ == "__main__":
    asyncio.run(main())

File: test_vs/05_agent_pipeline.rs
// Benchmark: Agent Pipeline Processing Messages — Rust baseline.
use std::sync::mpsc;
use std::thread;
use std::time::Instant;

fn main() {
    let (p_in_tx, p_in_rx) = mpsc::sync_channel::<String>(1024);
    let (p_out_tx, p_out_rx) = mpsc::sync_channel::<i64>(1024);
    let (v_out_tx, v_out_rx) = mpsc::sync_channel::<i64>(1024);
    let (t_out_tx, t_out_rx) = mpsc::sync_channel::<i64>(1024);

    // Parser
    thread::spawn(move || {
        while let Ok(raw) = p_in_rx.recv() {
            p_out_tx.send(raw.len() as i64).unwrap();
        }
    });

    // Validator
    thread::spawn(move || {
        while let Ok(value) = p_out_rx.recv() {
            v_out_tx.send(if value > 0 { value } else { 0 }).unwrap();
        }
    });

    // Transformer
    thread::spawn(move || {
        while let Ok(value) = v_out_rx.recv() {
            t_out_tx.send(value * 3 + 7).unwrap();
        }
    });

    let start = Instant::now();
    let mut total: i64 = 0;
    for i in 0..1000 {
        let msg = format!("message_payload_data_{}", i);
        p_in_tx.send(msg).unwrap();
        let result = t_out_rx.recv().unwrap();
        total += result;
    }
    drop(p_in_tx);
    let elapsed = start.elapsed().as_secs_f64();

    println!("{}", total);
    println!("__BENCH_METRICS__");
    println!("wall_time_s={:.6}", elapsed);
    println!("cpu_time_s={:.6}", elapsed);
    println!("peak_memory_kb=0");
}
